Chromatin activity identifies differential gene regulation across human ancestries
================

We have conducted the first genome-wide comparison of chromatin activity and 
contact frequency between human populations to identify non-protein-coding, 
or cis-regulatory elements (CREs) under recent selection. Our development of 
population-level ABC scores linking CREs to target genes and decomposition
into activity and contact components allowed us to find genes with ancestry-
associated differential expression across multiple contexts that could be 
identified by the differential activity of their enhancers at baseline. We 
further used our ancestry-associated differential activity CREs to find binding
differences in the transcription factors NF-\kappaB, JunD, and PU.1 that are 
condordant with allele frequency divergence and consistent with selection. As 
more information comes out linking DNA sequence to chromatin activity (e.g., 
bQTL) and the latter is assayed in more cell types from geographically and 
ancestrally diverse populations, the pipelines contained herein will be useful
for further study of selection has impacted modern disease phenotypes. 

Nextflow pipelines and scripts required for the analyses in our manuscript are 
contained within these directories. Required inputs are detailed and commented 
in the `.config` files. Detailed descriptions of all methods and analyses can 
be found in our preprint (2022).

